[Connects #125] Send messages to all users

mrgilman opened the pull request:
I reused the existing `TestMessagesController` and `MessageSender` classes, which required adding some conditionals, but reduces duplication.

mrgilman commented:
Should the `last_sent_at` be displayed in EST/EDT? Sent messages are displayed in the time zone of the recipient; that does not apply here.

jessieay commented on app/controllers/test_messages_controller.rb:
```diff
@@ -1,27 +1,39 @@
     employee = Employee.find_by(slack_username: slack_username)
 
-    if !employee.slack_channel_id.nil? || !employee.slack_user_id.nil?
+    if employee && (employee.slack_channel_id || employee.slack_user_id)
```
ah yes we have some definite improvement to do in this controller! (pre-your contributions)

jessieay commented on app/controllers/messages_controller.rb:
```diff
@@ -0,0 +1,27 @@
+      flash[:notice] = "Message created successfully"
+      redirect_to messages_path
+    else
+      flash.now[:error] = "Could not create message"
```
maybe include the actual message errors?

mrgilman replied:
Sure. The only validation is presence, and users won't be able to submit the form with blank fields, but I will add the error messages too just in case.

jessieay commented on app/controllers/messages_controller.rb:
```diff
@@ -0,0 +1,27 @@
+class MessagesController < ApplicationController
```
can we add some `before_action`s in here to ensure that only an admin can create a message?

mrgilman replied:
Yup, done

jessieay commented on app/controllers/send_messages_controller.rb:
```diff
@@ -0,0 +1,9 @@
+class SendMessagesController < ApplicationController
+  def create
```
can we add a `before_action` to ensure that only admins can send these?

mrgilman replied:
Ah, yup!

jessieay commented on app/controllers/test_messages_controller.rb:
```diff
@@ -1,27 +1,39 @@
 
   private
 
+  def message
```
is this method being used?

mrgilman replied:
Yup, line 16.

jessieay replied:
ok cool - should we also use it for `def new` ?

mrgilman replied:
Sure! Unfortunately we'll still need the conditional.

jessieay commented on spec/features/send_test_message_spec.rb:
```diff
@@ -14,8 +14,56 @@
+    fill_in "Slack username", with: username_from_fixture
+    click_on "Send test"
+
+    expect(page).to have_content("isn't up to date")
```
note to self: we should put flash messages in I18n so they are easier to update and check for in tests

jessieay commented on spec/features/send_message_spec.rb:
```diff
@@ -0,0 +1,32 @@
+require "rails_helper"
+
+feature "Send message" do
+  scenario "message sends successfully", js: true do
```
can we write a test that ensures a non-admin can't send a message to all employees? Want to make 100% sure non-admins can't spam the whole org ;)

mrgilman replied:
Sure thing. I'll add the `before_action` and a controller spec.

mrgilman commented:
Hi @jessieay, thanks for the thorough review! I think I've addressed all of your feedback, and all of the Code Climate errors other than the frozen strings. I've been trying to fix a [flaky Timecop test](https://github.com/18F/dolores-landingham-bot/pull/193/files#diff-7fd990b0fb4fafe3b912dff8e1a04c11R14) but I think the current implementation should fix it.

I'm heading out of the office in a few minutes, and am working on a different project tomorrow, but let me know if there is anything else I can do to finish this up, I'm sure I can find some time.

jessieay commented:
This all lgtm! Passing the baton to @harrisj

mrgilman commented:
I'll squash so this is ready to go on 👍 .

jessieay commented:
Just tested this locally and received the message 4 times...let me see if this is something on my end or an issue where the code needs to be updated

jessieay commented:
One error I am seeing locally:

The import script imports *all* users, even bot users, which is fine by me but when we try to send a message to all users I see this error:

> "cannot_dm_bot

thoughts on how we can skip DMing bot users when we send these?

jessieay commented:
I've confirmed that I am receiving the messages multiple times. 

I think this is because some of the jobs are failing due to the above bot dm issue. The job runs for *all* users and succeeds for me but fails when it tries to DM the bot. So it retries and DMs me again.

Perhaps in addition to fixing that we should send each message in its own DJ so people don't receive multiple messages if the job fails when sending to a different user.

jessieay commented:
ps - if you rebase now, those code climate errors should be fixed ;)

jessieay commented:
@mrgilman hello! don't want to pressure you, but any update on fixing the "dm to bot" issue?

mrgilman commented:
Hi @jessieay! Planning to get back to finishing this up in the next hour. Sorry it didn;t work perfectly the first time around!

jessieay commented:
@mrgilman no problem! that's why we do manual QA :) thanks for the update

mrgilman commented:
@jessieay Is there any reason to store the bot users in the DB? If not, I'm leaning towards skipping the import, rather than storing and flagging them.

jessieay commented:
@mrgilman can't think of a reason to store them.

mrgilman commented:
@jessieay I made the suggested changes, and also removed an unused class and spec that I left in my other PR. Let me know how this works!

Also, awesome to see a green Code Climate build! 💚

jessieay commented:
@mrgilman woo! looks great to me. will let @harrisj take a final look tonight. thanks so much for the updates.

harrisj commented:
This looks good to me. I think in the future we might want to change names since it's a bit confusing at first to understand the difference between `Message` and `ScheduledMessage` (maybe the first could called `BroadcastMessage`). But this isn't really a demerit against the PR. In addition, it looks like we need to use some i18n in Dolores, but also not in scope of this auction. Good work!

jessieay commented:
ooooh @harrisj I like that naming suggestion! Going to create an issue for it. `Message` was my suggestion but you're right, it's not very clear.
